### ISSUE TITLE
Agent fails to fetch sources when a task is re-using the GitHub service endpoint

### ISSUE THREAD

**Issue by jessehouwing** [Microsoft/vsts-agent#737]
## **Agent version and platform**
Version of your agent?
2.107.0.

OS of the machine running the agent? OSX/Windows/Linux...
Windows

## **VSTS type and version**
VisualStudio.com 
If VisualStudio.com, what is your account name? 
http://jessehouwing-dev.visualstudio.com

## **What's not working?** 
When I create a build task that uses the GitHub service endpoint type (in order to do further integration into GitHub), like this:
```
{
  "id": "123d700-cd1b-11e6-8ad3-3b209c1bb8f6",
  "name": "VSTSGithubReleaseTask",
  "friendlyName": "GitHub Release Task",
  "description": "Publish or update a release on Github",
  "author": "jessehouwing",
  "helpMarkDown": "",
  "category": "Deploy",
  "visibility": [
    "Build",
    "Release"
  ],
  "minimumAgentVersion": "1.95.0",
  "inputs": [
    {
      "name": "serviceEndpoint",
      "type": "connectedService:github",
      "label": "Github Account",
      "defaultValue": "",
      "required": true,
      "helpMarkDown": "Select or configure the service endpoint to provide your Github account."
    }
  ],
  "execution": {
    "Node": {
      "target": "vsts-github-release.js"
    }
  }
}
```
The task itself doesn't even get executed, adding the task and setting the service endpoint to any GitHub account (doesn't have to match the current repos endpoint) will cause the Sync Sources step to fail:

```
Syncing repository: 35dca4e7-a104-4c4b-984d-9284a042bf6f (GitHub)
Prepending Path environment variable with directory containing 'git.exe'.
git version
git init "C:\TfsData\Build-Dev-v2\_work\11\s"
Initialized empty Git repository in C:/TfsData/Build-Dev-v2/_work/11/s/.git/
git remote add origin http://github.com/
git config gc.auto 0
git config --get-all http.http://github.com/.extraheader
git remote set-url origin http://OAuth:********@github.com/
git remote set-url --push origin http://OAuth:********@github.com/
git fetch --tags --prune --progress origin
fatal: repository 'https://github.com/' not found
Git fetch failed with exit code: 128
```

I was hoping to give my users the ability to re-use the existing service endpoint, so they won't have to re-authenticate and so that I can piggy-back on the existing OAuth dialog implementation.

When I remove or disable the task, everything works as expected.

## **Agent and Worker's diag log**
Logs are located at the `_diag` under agent root folder, agent log prefix with `Agent_`, worker log prefix with `Worker_`. all sensitive information should already be masked out, please double check before paste here. 

[Worker_20161228-190707-utc.txt](https://github.com/Microsoft/vsts-agent/files/676134/Worker_20161228-190707-utc.txt)

Extension that shows this issue:
[jessehouwing.vsts-github-tasks-DEV-0.0.6.vsix.zip](https://github.com/Microsoft/vsts-agent/files/676141/jessehouwing.vsts-github-tasks-DEV-0.0.6.vsix.zip)

**Comment by ericsciple** [comment]
[this looks](https://github.com/Microsoft/vsts-agent/blob/1195c0f9858acf7deeb57bb4534236c8b130dec8/src/Agent.Worker/Build/BuildJobExtension.cs#L264) like the bug in the agent

**Comment by ericsciple** [comment]
it looks like a workaround (solution?) may be to prefer an endpoint where the $(build.repository.uri) matches the endpoint.data.repositoryId or endpoint.url.

or possibly even more simply, prefer endpoints where the repository.data.repositoryId is not null or empty.

**Comment by jessehouwing** [comment]
Can't do that, the uri is not configured through the Ui... It's just
hard-coded to http://github.com.

**Comment by ericsciple** [comment]
...sorry i mean for a fix in the agent.

**Comment by jessehouwing** [comment]
I get that. And thanks for thinking along with me. It's not possible to
'clone' the GitHub endpoint definition to my extension somehow, I
suspect... Right?

**Comment by ericsciple** [comment]
tasks have access to all endpoints sent down in the build job message. see the environment.endpoints section of the job message (big json structure dumped near the worker diag log). so you could access the repo endpoint (the one that is used to sync sources). is that what you mean? would that work, or does your task require different permissions?

**Comment by jessehouwing** [comment]
Since the Visual Studio endpoint is just about God on GitHub, it would
work, but I want to be able to select which account to use. And I want to
use it in rm as well. There it may just work, haven't tested yet. For now
let me try to link it to the build. That allows me to proceed.

I can wait though... For a real solution that is. Even if it limits the tfs
versions I can use the task with to 2017u1 and higher.

Can I demand on this or force a minimal agent version?

**Comment by ericsciple** [comment]
yes agent-version demand in the task.json should work. it looks like it will be a small fix and we should target getting it into the agent that 2017 U1 points to. folks on 2017 RTM (and *nix users on tfs 2015) would be able to download a newer agent.

**Comment by jessehouwing** [comment]
Sounds good :).

**Comment by jessehouwing** [comment]
The problem seems worse. Whenever I reference a Github endpoint in my build definition, it always tries to download the sources from there. Even if I setup the build definition to look at a VSTS Git repo and/or if I check the "don't sync sources box":

![image](https://cloud.githubusercontent.com/assets/4173387/21545989/09166598-cddc-11e6-9419-3b0293120e3a.png)

```
2016-12-29T14:30:52.0566274Z ##[section]Starting: Get Sources
2016-12-29T14:30:52.2686269Z Syncing repository: 35dca4e7-a104-4c4b-984d-9284a042bf6f (GitHub)
2016-12-29T14:30:52.2766257Z Prepending Path environment variable with directory containing 'git.exe'.
2016-12-29T14:30:52.2866269Z ##[command]git version
2016-12-29T14:30:52.4096259Z ##[command]git config --get remote.origin.url
2016-12-29T14:30:52.5246273Z ##[command]git config gc.auto 0
2016-12-29T14:30:52.6436265Z ##[command]git config --get-all http.http://github.com/.extraheader
2016-12-29T14:30:52.7256266Z ##[command]git remote set-url origin http://OAuth:********@github.com/
2016-12-29T14:30:52.8426278Z ##[command]git remote set-url --push origin http://OAuth:********@github.com/
2016-12-29T14:30:52.9276276Z ##[command]git fetch --tags --prune --progress origin
2016-12-29T14:30:54.0079642Z fatal: repository 'https://github.com/' not found
2016-12-29T14:30:54.1229622Z ##[error]Git fetch failed with exit code: 128
2016-12-29T14:30:54.1339645Z ##[section]Finishing: Get Sources
```

**Comment by jessehouwing** [comment]
Also, it looks like the default URI of the Git endpoint is set to be insecure `http://github.com`, which may cause issues later when I try to fetch repo's and accounts, releases etc to populate dropdowns in the UI.

**Comment by jessehouwing** [comment]
I've been experimenting with extending other people's service endpoints with more datasources, and I can't find the syntax. As part of this, I'd love it if we could define additional datasources for an existing endpoint. This would allow me to query things VSTS is not yet doing.

**Comment by TingluoHuang** [comment]
close as the agent that contains the fix has rolls out.

**Comment by jessehouwing** [comment]
Seems to be fixed with agent: 2.112.0.

Thanks, this will allow me to proceed with what I was trying to accomplish.
